test/index.ts

```diff
@@ -249,4 +271,99 @@ describe('teeny', () => {
       });
     });
   });
+
+  it('should track stats, callback mode, success', done => {
+    const scope = mockJson();
+    teenyRequest({uri}, () => {
+      assert.ok(statsInstanceStubs.requestStarting.calledOnceWithExactly());
+      assert.ok(statsInstanceStubs.requestFinished.calledOnceWithExactly());
+      scope.done();
+      done();
+    });
+  });
+
+  it('should track stats, callback mode, failure', done => {
+    const scope = mockError();
+    teenyRequest({uri}, err => {
+      assert.ok(err);
+      assert.ok(statsInstanceStubs.requestStarting.calledOnceWithExactly());
+      assert.ok(statsInstanceStubs.requestFinished.calledOnceWithExactly());
+      scope.done();
+      done();
+    });
+  });
+
+  it('should track stats, stream mode, success', done => {
+    const scope = mockJson();
+    const readable = teenyRequest({uri});
+    assert.ok(statsInstanceStubs.requestStarting.calledOnceWithExactly());
+
+    readable.once('response', () => {
+      assert.ok(statsInstanceStubs.requestFinished.calledOnceWithExactly());
+      scope.done();
+      done();
+    });
+  });
+
+  it('should track stats, stream mode, failure', done => {
+    const scope = mockError();
+    const readable = teenyRequest({uri});
+    assert.ok(statsInstanceStubs.requestStarting.calledOnceWithExactly());
+
+    readable.once('error', err => {
+      assert.ok(err);
+      assert.ok(statsInstanceStubs.requestFinished.calledOnceWithExactly());
+      scope.done();
+      done();
+    });
+  });
+
+  // TODO multipart is broken with 2 strings
+  it.skip('should track stats, multipart mode, success', done => {
+    const scope = mockJson();
+    teenyRequest(
+      {
+        headers: {},
+        multipart: [{body: 'foo'}, {body: 'bar'}],
+        uri,
+      },
+      () => {
+        assert.ok(statsInstanceStubs.requestStarting.calledOnceWithExactly());
+        assert.ok(statsInstanceStubs.requestFinished.calledOnceWithExactly());
+        scope.done();
+        done();
+      }
+    );
+  });
+
+  it.skip('should track stats, multipart mode, failure', done => {
+    const scope = mockError();
+    teenyRequest(
+      {
+        headers: {},
+        multipart: [{body: 'foo'}, {body: 'bar'}],
+        uri,
+      },
+      err => {
+        assert.ok(err);
+        assert.ok(statsInstanceStubs.requestStarting.calledOnceWithExactly());
+        assert.ok(statsInstanceStubs.requestFinished.calledOnceWithExactly());
+        scope.done();
+        done();
+      }
+    );
+  });
+
+  it('should pass teeny statistics options', () => {
+    const opts = {concurrentRequests: 42};
+    teenyRequest.setStatOptions(Object.assign({}, opts));
+    assert(statsInstanceStubs.setOptions.calledOnceWithExactly(opts));
+  });
+
+  it('should return teeny statistics options', () => {
```

Comment:
I'd be tempted to add one test here which confirms that `process.on('warning')` is emitted if you create too many concurrent requests (you can just set the threshold to 1 for this test).

Reply:
Since `process.on('warning')` is handled by `TeenyStatistics` and not the `teenyRequest` function itself, this would be an integration test. Despite that, I agree this test should exist. I ended up refactoring the integration code of `teenyRequest` a bit to facilitate this type of test. As a result the following was changed relative to the initial code in this PR:

- much cleaner and simplified stubbing
- removed the dependency on proxyquire
- exposed `TeenyStatistics` as `stats`, and no pass-through methods are required
- added the ability to reset statistics
- added `TeenyStatistics.getOptions` (i opted for `getOptions`/`setOptions` over real getters/setters)

overall a much cleaner result, thanks for the suggestion!